core/src/main/kotlin/org/evomaster/core/output/dto/DtoWriter.kt

```diff
@@ -56,29 +56,66 @@ class DtoWriter {
         actionDefinitions.forEach { action ->
             action.getViewOfChildren().find { it is BodyParam }
             .let {
-                val primaryGene = (it as BodyParam).primaryGene().getLeafGene()
-                // TODO: Payloads could also be json arrays, analyze ArrayGene
-                when (primaryGene) {
-                    is ObjectGene -> {
-                        calculateDtoFromObject(primaryGene, action.getName())
-                    }
-                    is ArrayGene<*> -> {
-                        calculateDtoFromArray(primaryGene, action.getName())
-                    }
-                    else -> {
-                        throw IllegalStateException("Gene $primaryGene is not supported for DTO payloads")
+                val primaryGene = (it as BodyParam).primaryGene()
+                if (primaryGene is ChoiceGene<*>) {
+                    calculateDtoFromChoice(primaryGene, action.getName())
+                } else {
+                    calculateDtoFromNonChoiceGene(primaryGene.getLeafGene(), action.getName())
+                }
+            }
+        }
+    }
+
+    private fun calculateDtoFromChoice(gene: ChoiceGene<*>, actionName: String) {
+        val dtoName = TestWriterUtils.safeVariableName(actionName)
+        val dtoClass = DtoClass(dtoName)
+        val children = gene.getViewOfChildren()
+        // merge options into a single DTO
+        children.forEach { childGene ->
+            when (childGene) {
+                is ObjectGene -> populateDtoClass(dtoClass, childGene)
+                is ArrayGene<*> -> {
+                    val template = childGene.template
+                    if (template is ObjectGene) {
+                        populateDtoClass(dtoClass, template)
                     }
                 }
             }
         }
+        dtoCollector.put(dtoName, dtoClass)
+    }
+
+    private fun calculateDtoFromNonChoiceGene(gene: Gene, actionName: String) {
+        when (gene) {
+            is ObjectGene -> calculateDtoFromObject(gene, actionName)
+            is ArrayGene<*> -> calculateDtoFromArray(gene, actionName)
+            else -> {
+                throw IllegalStateException("Gene $gene is not supported for DTO payloads")
```

Comment:
add some more info to help debugging if it happens, eg put the `actionName` somewhere in this error message

Reply:
Added! This exception throwing should actually be short lived, I was thinking it's main purpose will be providing info for when we execute against Dataset to see which cases we're missing